### PR TITLE
Implement short link generation with redirect

### PR DESCRIPTION
**Pull request by AnotherDevvv**

### CODE REVIEW

File: internal/api/shortener.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	}
+}
+
+func (sh *Shortener) ShortenUrl(c echo.Context) error {
+	url := c.QueryParam("url")
+
+	fmt.Println(url)
+
+	e := validate(url)
+	if e != nil {
```
Comment (LasTshaMAN) [inline internal/api/shortener.go:32]: usually `err` is used as error variable when possible

File: internal/db/db.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+type UrlRepository struct {
+	db      * bolt.DB
+}
+
+func NewUrlRepository() *UrlRepository {
+	urlRep := new(UrlRepository)
+
+	db, err := bolt.Open("shortener.db", 0600, nil)
+
+	if err != nil {
```
Comment (LasTshaMAN) [inline internal/db/db.go:25]: usually errors are handled on the very next line after function execution, so there is no need for a blank line in between,

this "convention" helps a bit with readability

File: internal/db/db.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+
+func (urlRep *UrlRepository) Close() error {
+	return urlRep.db.Close()
+}
+
+func (urlRep *UrlRepository) Insert(key []byte, url string) error {
+	err := urlRep.db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte("shortener"))
```
Comment (LasTshaMAN) [inline internal/db/db.go:41]: bucket name is duplicated, maybe define a const for it

File: internal/db/db.go (hunk 2 above)
Comment (LasTshaMAN) [reply internal/db/db.go:25]: This is applicable to other places in this PR

File: internal/db/db.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+		if err != nil {
+			return fmt.Errorf("create bucket: %s", err)
+		}
+
+		if b.Put(key, []byte(url)) != nil {
+			return err
```
Comment (LasTshaMAN) [inline internal/db/db.go:48]: its a good practice to wrap errors with `fmt.Errorf("additional context: %w", err)` because there are no stack-traces in Go yet,

so we kinda have to add stack-traces ourselves (usually this is done when it might not be clear which `func` error originates from)

File: internal/db/db.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		}
+		return nil
+	})
+
+	return err
+}
+
+func (urlRep *UrlRepository) Get(key []byte) (string, error) {
+	v := []byte{}
```
Comment (LasTshaMAN) [inline internal/db/db.go:57]: image

File: internal/db/db.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+	err := urlRep.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("shortener"))
+
+		k := b.Get(key)
```
Comment (LasTshaMAN) [inline internal/db/db.go:62]: maybe use a bit more descriptive names, might not be clear what `k` represents

File: internal/api/shortener_test.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	hash := sha256.Sum256([]byte(url))
+	req, err := http.NewRequest(
+		echo.POST,
+		fmt.Sprintf("%sshorten?url=%s", address, url),
+		strings.NewReader(``),
+	)
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		t.Fatal(err)
```
Comment (LasTshaMAN) [inline internal/api/shortener_test.go:61]: could use `require.NoError` instead

File: internal/api/shortener_test.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+func TestMain(m *testing.M) {
+	setUp()
+
+	retCode := m.Run()
+
+	os.Exit(retCode)
+}
+
+var e *echo.Echo
+var sh *Shortener
```
Comment (LasTshaMAN) [inline internal/api/shortener_test.go:30]: global variables are discouraged (even in tests, because over time tests might become large and unmanageable as well), try to get rid of these

File: internal/api/shortener.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+		return c.String(http.StatusBadRequest, fmt.Sprintf("URL cannot be shortened because %s", e.Error()))
+	}
+
+	hash := sha256.Sum256([]byte(url))
+
+	err := sh.urlRepository.Insert(hash[:4], url)
```
Comment (LasTshaMAN) [inline internal/api/shortener.go:38]: better have some comment about `hash[:4]`, why this slice gets truncated, and why do we even use hashes as keys (instead of URLs)

**Review by LasTshaMAN — COMMENTED** [review]
Mostly syntax (and code-structure) related comments for now,

will take another look once these are resolved :)

File: internal/api/shortener.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+	if err != nil {
+		return c.String(http.StatusServiceUnavailable, e.Error())
+	}
+
+	return c.String(http.StatusOK, sh.address + hex.EncodeToString(hash[:4]))
+}
+
+func (sh *Shortener) FollowUrl(c echo.Context) error  {
```
Comment (LasTshaMAN) [inline internal/api/shortener.go:47]: you might wanna cover `FollowUrl` with tests as well,

also, would be nice to have a test that gets a short URL version via `ShortenUrl` and then gets redirected where expected after calling `FollowUrl`,
might do a table-test (with a bunch of different URL examples in it) as well to check some edge-cases

File: internal/api/shortener.go (hunk 9 above)
Comment (LasTshaMAN) [reply internal/api/shortener.go:38]: Also, it seems `hash[:4]` is duplicated (for insert, then for fetch), so maybe create a separate func that converts URL to key, and leave some comment description there

**Comment by LasTshaMAN** [comment]
Wound be nice to document some general assumptions this service operates under (in README)

File: cmd/api/main.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"fmt"
+	"github.com/labstack/echo/v4"
+	"urlShortener/internal/api"
+	"urlShortener/internal/db"
+)
+
+const port = "1323"
+
+func main() {
```
Comment (LasTshaMAN) [inline cmd/api/main.go:12]: would be nice to add interrupt signal handler to `main` so that program can be terminated gracefully

File: internal/db/db.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	"fmt"
+	"github.com/boltdb/bolt"
+)
+
+//go:generate moq -pkg mock -out mock/repository.go . Repository
+
+type Repository interface {
+	Insert(key []byte, url string) error
+	Get(key []byte) (string, error)
+	Close() error
```
[the remark this answers is not in the document]
Comment (AnotherDevvv) [reply internal/db/db.go]: deferred at main. I expect it to close resources gracefully during signal handler termination process

File: internal/db/db.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"fmt"
+	"github.com/boltdb/bolt"
+)
+
+//go:generate moq -pkg mock -out mock/repository.go . Repository
+
+type Repository interface {
+	Insert(key []byte, url string) error
```
[the remark this answers is not in the document]
Comment (LasTshaMAN) [reply internal/db/db.go]: yes, doesn't matter whether its a prefix or suffix, acronyms are usually UPPERCASE in Go

File: cmd/api/main.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	router := api.NewRouter(shortener)
+	go router.Start()
+
+	awaitTermination(router, repository)
+}
+
+func initLogger()  {
+	logfile, err := os.OpenFile("logs", os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		panic("Unable to create log file")
```
Comment (LasTshaMAN) [inline cmd/api/main.go:29]: using `panic`s is discouraged, better to return error from `initLogger` and handle it in `main.go` by `os.Exit(1)`

File: internal/api/shortener.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+	log.Debugf("Shortening %s", url)
+
+	err := validate(url)
+	if err != nil {
+		return c.String(http.StatusBadRequest, fmt.Sprintf("URL cannot be shortened because %s", err))
+	}
+
+	key := encode(url)
+	value, err := sh.urlRepository.Get(key)
```
Comment (LasTshaMAN) [inline internal/api/shortener.go:36]: `err` is not handled, linters (or IDE) ideally should catch that

File: internal/api/shortener_test.go [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+	serverFailURL = "http://hostname/failme"
+)
+
+func TestShortener_ShortenURL(t *testing.T) {
+	testCases := []struct {
+		url              string
+		expectedStatus   int
+		expectedCode     string
+		expectedLocation string
+		repository       db.Repository
```
Comment (LasTshaMAN) [inline internal/api/shortener_test.go:30]: usually `testName` (or similar) parameter is added so that it is clear which test fails

File: internal/api/shortener_test.go [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+			fmt.Sprintf("%sshorten?url=%s", host, tc.url),
+			strings.NewReader(``),
+		)
+
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+
+
```
Comment (LasTshaMAN) [inline internal/api/shortener_test.go:100]: no need to use double new-lines

File: internal/api/shortener_test.go [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+			defer resp.Body.Close()
+
+			require.Equal(t, http.StatusMovedPermanently, resp.StatusCode)
+			require.Equal(t, tc.url, resp.Header.Get("Location"))
+		}
+
+		err = closer()
+		if err != nil {
+			continue
+		}
```
Comment (LasTshaMAN) [inline internal/api/shortener_test.go:122]: just assert that err is nil in tests

File: internal/db/db.go [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+	db *bolt.DB
+}
+
+func NewURLRepository() *URLRepository {
+	return &URLRepository{
+		db: func() *bolt.DB {
+			db, err := bolt.Open("shortener.db", 0600, nil)
+			if err != nil {
+				log.Fatalf("Unable to create embedded db due to %s", err)
+				panic(err)
```
Comment (LasTshaMAN) [inline internal/db/db.go:28]: better not use panic, and return errors explicitely

File: internal/db/db.go [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+			return db
+		}(),
+	}
+}
+
+func (urlRep *URLRepository) Close() error {
+	err := urlRep.db.Close()
+	if err != nil {
+		return fmt.Errorf("Failed to close db gracefully %w", err)
```
Comment (LasTshaMAN) [inline internal/db/db.go:39]: `Failed to close db gracefully %w`->`failed to close db gracefully: %w`

Error message should start with lower-case letter because it might be wrapped with another error (and output will be a bit ugly)

File: internal/db/db.go [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+type URLRepository struct {
+	db *bolt.DB
+}
+
+func NewURLRepository() *URLRepository {
+	return &URLRepository{
+		db: func() *bolt.DB {
+			db, err := bolt.Open("shortener.db", 0600, nil)
+			if err != nil {
+				log.Fatalf("Unable to create embedded db due to %s", err)
```
Comment (LasTshaMAN) [inline internal/db/db.go:27]: `log.Fatalf` will call panic as well due to its implementation, if I remember correctly

File: internal/api/router.go [hunk 22]
```diff
@@ -4,13 +4,10 @@ import (
 	stdContext "context"
```
Comment (LasTshaMAN) [inline internal/api/router.go]: no need to add `stdContext` here
